File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] result = new int[2];
        result[0] = -1; 
        result[1] = -1;

        // find first occurrence
        int i = 0, j = nums.length - 1;
        while (i <= j) {
            int mid = i + (j - i) / 2;
            if (nums[mid] >= target) {
                j = mid - 1;
            } else {
                i = mid + 1;
            }
            if (nums[mid] == target) {
                result[0] = mid;
            }
        }

        // find last occurrence
        i = 0; j = nums.length - 1;
        while (i <= j) {
            int mid = i + (j - i) / 2;
            if (nums[mid] <= target) {
                i = mid + 1;
            } else {
                j = mid - 1;
            }
            if (nums[mid] == target) {
                result[1] = mid;
            }
        }

        return result;
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution(object):
    def searchRange(self, nums, target):
        def findLeft():
            i, j = 0, len(nums) - 1
            left = -1
            while i <= j:
                mid = (i + j) // 2
                if nums[mid] >= target:
                    j = mid - 1
                else:
                    i = mid + 1
                if nums[mid] == target:
                    left = mid
            return left
        
        def findRight():
            i, j = 0, len(nums) - 1
            right = -1
            while i <= j:
                mid = (i + j) // 2
                if nums[mid] <= target:
                    i = mid + 1
                else:
                    j = mid - 1
                if nums[mid] == target:
                    right = mid
            return right

        return [findLeft(), findRight()]
